File: java/src/test/java/ai/buble/sdk/ChatTest.java
package ai.buble.sdk;

import ai.buble.sdk.streaming.BubleStream;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

final class ChatTest {
    @Test
    void streamsOpenAIText() throws Exception {
        String sse = "data: {\"choices\":[{\"delta\":{\"content\":\"Hel\"}}]}\n\n"
                + "data: {\"choices\":[{\"delta\":{\"content\":\"lo\"}}]}\n\n"
                + "data: [DONE]\n\n";
        try (TestHttpServer server = new TestHttpServer(request ->
                new TestHttpServer.Response(200, sse).header("Content-Type", "text/event-stream"))) {
            BubleClient client = BubleClient.builder().apiKey("sk_test").baseUrl(server.url()).build();

            StringBuilder out = new StringBuilder();
            try (BubleStream stream = client.chat().completions().stream(Map.of(
                    "model", "openai/gpt-5.5",
                    "messages", List.of(Map.of("role", "user", "content", "hello"))
            ))) {
                while (stream.next()) {
                    out.append(stream.text());
                }
            }

            assertEquals("Hello", out.toString());
            assertEquals("/api/v1/chat/completions", server.requests().get(0).path);
            assertTrue(server.requests().get(0).body.contains("\"stream\":true"));
        }
    }

    @Test
    void encodesGeminiModelPath() throws Exception {
        try (TestHttpServer server = new TestHttpServer(request ->
                new TestHttpServer.Response(200, "{\"candidates\":[]}"))) {
            BubleClient client = BubleClient.builder().apiKey("sk_test").baseUrl(server.url()).build();

            client.chat().gemini().generateContent("google/gemini 2.5", Map.of("contents", List.of()));

            assertEquals("/api/v1beta/models/google/gemini%202.5:generateContent", server.requests().get(0).path);
        }
    }
}

File: rust/tests/chat_test.rs
use buble::Client;
use futures_util::StreamExt;
use serde_json::json;
use wiremock::{
    matchers::{body_json, method, path},
    Mock, MockServer, ResponseTemplate,
};

#[tokio::test]
async fn creates_openai_chat_completion() {
    let server = MockServer::start().await;
    Mock::given(method("POST"))
        .and(path("/api/v1/chat/completions"))
        .and(body_json(json!({
            "model": "openai/gpt-5.4",
            "messages": [{ "role": "user", "content": "Hi" }],
            "stream": false
        })))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "choices": [{ "message": { "content": "Hello" } }]
        })))
        .mount(&server)
        .await;

    let response = client(&server)
        .chat()
        .completions()
        .create(json!({
            "model": "openai/gpt-5.4",
            "messages": [{ "role": "user", "content": "Hi" }]
        }))
        .await
        .expect("chat");

    assert_eq!(
        response.pointer("/choices/0/message/content").unwrap(),
        "Hello"
    );
}

#[tokio::test]
async fn streams_openai_text() {
    let server = MockServer::start().await;
    Mock::given(method("POST"))
        .and(path("/api/v1/chat/completions"))
        .respond_with(
            ResponseTemplate::new(200)
                .insert_header("content-type", "text/event-stream")
                .set_body_string("data: {\"choices\":[{\"delta\":{\"content\":\"Hel\"}}]}\n\ndata: {\"choices\":[{\"delta\":{\"content\":\"lo\"}}]}\n\ndata: [DONE]\n\n"),
        )
        .mount(&server)
        .await;

    let mut stream = client(&server)
        .chat()
        .completions()
        .stream_text(json!({
            "model": "openai/gpt-5.4",
            "messages": [{ "role": "user", "content": "Hi" }]
        }))
        .await
        .expect("stream");

    let mut output = String::new();
    while let Some(chunk) = stream.next().await {
        output.push_str(&chunk.expect("chunk"));
    }

    assert_eq!(output, "Hello");
}

#[tokio::test]
async fn calls_gemini_model_path_without_escaping_slash() {
    let server = MockServer::start().await;
    Mock::given(method("POST"))
        .and(path(
            "/api/v1beta/models/openai/gpt%2D5%2E4:generateContent",
        ))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({ "candidates": [] })))
        .mount(&server)
        .await;

    client(&server)
        .chat()
        .gemini()
        .generate_content("openai/gpt-5.4", json!({ "contents": [] }))
        .await
        .expect("gemini");
}

fn client(server: &MockServer) -> Client {
    Client::builder()
        .api_key("sk_test")
        .base_url(server.uri())
        .build()
        .expect("client")
}
